src/dataframe/dataframe.jl

```diff
@@ -1278,12 +1334,19 @@ and columns are matched by order of appearance. In this case `row` must contain
 the same number of elements as the number of columns in `df`.
 
 If `row` is a `DataFrameRow`, `NamedTuple` or `AbstractDict` then
-values in `row` are matched to columns in `df` based on names (order is ignored).
-`row` may contain more columns than `df` if `columns=:intersect`
-(this is currently the default, but will change in the future), but all column names
-that are present in `df` must be present in `row`.
-Otherwise if `columns=:equal` then `row` must contain exactly the same columns as `df`
-(but possibly in a different order).
+values in `row` are matched to columns in `df` based on names. The exact behavior
+depends on `cols` argument value in the following way:
+* If `cols=:equal` (this is currently the default, but will change to `:identical` in the future)
+  then `row` must contain exactly the same columns as `df` (but possibly in a different order).
+* If `cols=:identical` then `row` must contain the same columns in the same order
+  (this option is the same as `:equal` if `row` is an `AbstractDict`).
+* If `cols=:intersect` then `row` may contain more columns than `df`,
+  but all column names that are present in `df` must be present in `row`.
+* If `cols=:subset` then `push!` behaves like for `:intersect` but if some column
+  is missing in `row` then a `missing` value is pushed to `df`.
+* Finally, if `cols=:union` then `push!` behaves like for `:subset` with two additional changes:
+  a) if `row` contains columns that are not present in `df` then they are added to `df`
+  at the end and filled with `missing` for all rows existing in `df` prior to calling `push!.
```

Comment:
Though not related to this special pull request, besides union, all other symbols, especially `:intersect`, should mean the same for `push!` and `vcat` in my opinion.

Reply:
Isn't that the case overall? One difference is that `push!` doesn't promote column types, but at least that's explicit. We can discuss that elsewhere if needed.

Reply:
:intersect works differently for vcat and push!. The latter requires all column names of the dataframe to exist in the new row. The former builds the real intersection of columns and merges those.

Reply:
Ah, right. But then you're pushing to an existing data frame so it sounds legitimate that existing columns are preserved. It would be really weird for `push!` to remove columns.

Reply:
Hmm I feel like I see that whole problem from the wrong perspective: I can enter :intersect into push! thus the function should be capable to do the intersection. Well, if intersection is only supported, if the new row is an actual superset of columns I'd somehow not call it :intersect though it technically still is one.

Reply:
Thus I'd prefer to not support `:intersect` since with that restriction it is the same as `:subset` as far as I understand.

Reply:
`:intersect` and `:subset` for `push!` are not the same (the latter allows list of columns to be a strict subset and the former does not allow this). Can you please comment here or in https://github.com/JuliaData/DataFrames.jl/issues/1991 if you see a name that is better fit as a replacement of `:intersect`. Thank you!

Reply:
Leaving this for a separate PR.